guide/src/dev/performance.md

```diff
@@ -0,0 +1,206 @@
+    let result = _mm_cmpeq_epi8(tmp, bitmask);
+    _mm_movemask_epi8(result)
+}
+```
+
+This SIMD optimization can provide significant speedups when processing large documents, since character scanning is such a common operation. The code falls back to scalar processing when SIMD is not available.
+
+## Memory-Efficient String Storage
+
+The library uses a custom string type `CowStr` that can represent strings in three ways:
+
+1. `Borrowed` - Zero-copy reference to input text
+2. `Inlined` - Small strings stored directly in the enum
+3. `Boxed` - Heap allocated for larger strings
+
+```rust
+pub enum CowStr<'a> {
+    Boxed(Box<str>),
+    Borrowed(&'a str),
+    Inlined(InlineStr),
+}
+```
+
+This allows efficient handling of strings with different sizes and lifetimes:
+
+- Small strings can avoid allocation by using the inline storage
+- References to input text avoid copying when possible
+- Heap allocation is only used when necessary for large strings
+
+The `InlineStr` type can store strings up to 22 bytes (on 64-bit systems) directly in the enum variant, avoiding heap allocation for common short string cases. Refer to the [string handling](./string-handling.md) documentation for more details.
+
+## Tree Structure Optimization
+
+The AST (Abstract Syntax Tree) is stored in a vec-based tree structure that provides:
+
+1. Fast node creation during parsing
+2. Efficient tree traversal
+3. Memory locality from vector storage
+
+```rust
+pub(crate) struct Tree<T> {
+    nodes: Vec<Node<T>>,
+    spine: Vec<TreeIndex>,
+    cur: Option<TreeIndex>, 
+}
+
+pub(crate) struct Node<T> {
+    pub child: Option<TreeIndex>,
+    pub next: Option<TreeIndex>,
+    pub item: T,
+}
+```
+
+Key optimizations in the tree structure include:
+
+- Using indices instead of pointers for node references
+- Maintaining a "spine" for fast access to ancestor nodes
+- Storing nodes contiguously in a vector for better cache usage
+- Using non-zero indices to save space in option types
+
+## Single-Pass Parsing
+
+The parser uses a single pass over the input text for both block and inline parsing, avoiding the need to rescan text multiple times. This is achieved through:
+
+1. Maintaining parser state to track context
+2. Handling both block and inline elements incrementally
+3. Using lookahead only when necessary
+
+For example, here's how link reference definitions are handled in a single pass:
+
+```rust
+fn parse_refdef_total(&mut self, start: usize) -> Option<(usize, LinkLabel<'a>, LinkDef<'a>)> {
+    let bytes = &self.text.as_bytes()[start..];
+    if scan_ch(bytes, b'[') == 0 {
+        return None;
+    }
+    let (mut i, label) = self.parse_refdef_label(start + 1)?;
+    i += 1;
+    if scan_ch(&bytes[i..], b':') == 0 {
+        return None;
+    }
+    i += 1;
+    let (bytecount, link_def) = self.scan_refdef(start, start + i)?;
+    Some((bytecount + i, UniCase::new(label), link_def))
+}
+```
+
+## Lazy Text Merging
```

Comment:
This is not a performance optimization. It is to work around the performance optimization that Text events are often split into multiple parts for various reasons. (Zero-copy ranges may be disjoint & parsed special characters are cheaper to keep in `InlineStr` rather than concatenating into a new allocation.)

You may want to use `TextMergeStream` if you'd rather handle the consecutive events "merged" together, but it has a performance cost.